neqo-qpack/src/table.rs

```diff
@@ -192,51 +264,66 @@ impl HeaderTable {
         }
         self.base += 1;
         self.used += entry.size();
+        let index = entry.index();
         self.dynamic.push_front(entry);
-        Ok(())
+        Ok(index)
     }
 
     pub fn insert_with_name_ref(
         &mut self,
         name_static_table: bool,
         name_index: u64,
-        value: Vec<u8>,
+        value: &[u8],
     ) -> Res<()> {
+        qtrace!(
+            [self],
+            "insert with ref to index={} in {} value={:?}",
+            name_index,
+            if name_static_table {
+                "static"
+            } else {
+                "dynamic"
+            },
+            value
+        );
         let name;
         if name_static_table {
-            let entry = self.get_static(name_index)?;
-            name = entry.name().to_vec()
+            name = self.get_static(name_index)?.name().to_vec();
         } else {
-            let entry = self.get_dynamic(name_index, self.base, false)?;
-            name = entry.name().to_vec();
+            name = self
+                .get_dynamic(name_index, self.base, false)?
+                .name()
+                .to_vec();
         }
-        self.insert(name, value)
+        self.insert(&name, value)?;
+        Ok(())
     }
 
     pub fn duplicate(&mut self, index: u64) -> Res<()> {
-        // need to remember name and value because inser may delete the entry.
+        qtrace!([self], "dumplicate entry={}", index);
+        // need to remember name and value because insert may delete the entry.
         let name: Vec<u8>;
         let value: Vec<u8>;
         {
             let entry = self.get_dynamic(index, self.base, false)?;
             name = entry.name().to_vec();
             value = entry.value().to_vec();
+            qtrace!([self], "dumplicate name={:?} value={:?}", name, value);
         }
-        self.insert(name, value)?;
+        self.insert(&name, &value)?;
         Ok(())
     }
 
-    pub fn increment_acked(&mut self, increment: u64) {
+    pub fn increment_acked(&mut self, increment: u64) -> Res<()> {
+        qtrace!([self], "increment acked by {}", increment);
         self.acked_inserts_cnt += increment;
+        if self.base < self.acked_inserts_cnt {
+            return Err(Error::InternalError);
+        }
```

Comment:
could this be an assert or debug_assert? This would simplify signature of this fn as well as caling fn.

Reply:
no this function is called when an Insert Count Increment instruction is received. Server implementation can have bugs, therefore we should not crash.

Reply:
ah ok!